### PR TITLE
post: revoke apple id

### PR DESCRIPTION
**Pull request by nachoSource**

### CODE REVIEW

File: posts/Nacho-source/01.Revoking-Apple-Token.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+> The following steps are executed in your server, so in this case we are using [_Node.js_](https://nodejs.org/en/)
+
+The second JWT that we need is the `clientSecret`, which can be generated using a [_series of keys_](https://github.com/jooyoungho/apple-token-revoke-in-firebase/issues/1#issuecomment-1163098500) associated with your app.
+
+```javascript
+const jwt = require('jsonwebtoken');
+const fs = require('fs');
+
+const makeJWT = () => {
```
Comment (navarroaxel) [inline posts/Nacho-source/01.Revoking-Apple-Token.md:76]:
```suggestion
const makeJWT = async () => {
```

File: posts/Nacho-source/01.Revoking-Apple-Token.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
+---
+title: Revoking Apple ID
+published: false
```
Comment (navarroaxel) [inline posts/Nacho-source/01.Revoking-Apple-Token.md:3]: Use `true` to publish when is merged
```suggestion
published: true
```

File: posts/Nacho-source/01.Revoking-Apple-Token.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+description: Revoke Apple ID in order to pass app versions review for iOS
+tags: 'ios, mobile, javascript, reactnative'
+cover_image: ./assets/app-store.jpeg
+---
+
+In the past few weeks, an issue has been causing some headaches amongst the mobile developers community. By the end of May, [Apple has announced](https://developer.apple.com/news/?id=12m75xbj) that every new version of apps published in App Store that allow users to sign in through Apple IDs also must allow them to cancel their subscription through the use of a [specific API](https://developer.apple.com/documentation/sign_in_with_apple/revoke_tokens). The app versions that don't implement this requieremente will neither pass the review nor be available for use.
```
Comment (pabloromeo) [inline posts/Nacho-source/01.Revoking-Apple-Token.md:9]:
```suggestion
In the past few weeks, an issue has been causing some headaches amongst the mobile developers community. By the end of May, [Apple has announced](https://developer.apple.com/news/?id=12m75xbj) that every new version of apps published in App Store that allow users to sign in through Apple IDs also must allow them to cancel their subscription through the use of a [specific API](https://developer.apple.com/documentation/sign_in_with_apple/revoke_tokens). The app versions that don't implement this requirement will neither pass the review nor be available for use.
```

File: posts/Nacho-source/01.Revoking-Apple-Token.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+So far this seems pretty legit. The problem is that the documentation provided is, at least, insufficient when not confusing, and the official deadline for this was on June 30th. This pushed the whole community to develop different workarounds in order to meet this deadline.
+
+In this post I am sharing a solution that gathers conclusions extracted from the work of many developers.
+
+> _Disclaimer_
+> ❗Part of the problem is that the use of this API does not guarantee the effective disassociation between your app and the Apple ID -which is logically the pursued objective, even when it returns _200_ (and no other significant info). Leaving that aside, this workaround does the job as it satisfies the required behavior. The rest is Apple’s matter.
```
Comment (pabloromeo) [inline posts/Nacho-source/01.Revoking-Apple-Token.md:16]:
```suggestion
> ❗Part of the problem is that the use of this API does not guarantee the effective disassociation between your app and the Apple ID -which is logically the pursued objective, even when it returns _200_ (and no other significant info). Leaving that aside, this workaround does the job as it satisfies the required behavior. The rest is Apple’s problem.
```

File: posts/Nacho-source/01.Revoking-Apple-Token.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+## Table of contents:
+
+1. Saving the credential token
+2. Generating the `clientSecret`
+3. Consuming the API
+4. Conclusion
+
+### 1. Saving the credential token
+
+Although I am using [_React Native_](https://reactnative.dev), this solution should work globally if you can access the `identityToken` every time the user enters the app, either signing or loging in.
```
Comment (pabloromeo) [inline posts/Nacho-source/01.Revoking-Apple-Token.md:27]:
```suggestion
Although I am using [_React Native_](https://reactnative.dev), this solution should work globally if you can access the `identityToken` every time the user enters the app, either signing or logging in.
```

**Review by pabloromeo — CHANGES_REQUESTED** [review]

**Review by pabloromeo — APPROVED** [review]

File: posts/nachoSource/01.revoking-apple-token.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+```
+
+> :bulb: `fetch` is a global API since _Node.js_ v18 , so there's no need to import it here.
+
+In case you want or need the use of `authorizationCode`, you should add an extra step in order to get the `refreshToken`:
+
+```javascript
+const revokeToken = async (authorizationCode) => {
+    const fetch = require("node-fetch");
+
```
Comment (navarroaxel) [inline posts/nachoSource/01.revoking-apple-token.md:126]:
```suggestion
```

File: posts/nachoSource/01.revoking-apple-token.md [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+            method: "POST",
+            body: JSON.stringify({
+                client_id: parameter.clientId,
+                client_secret: clientSecret,
+                code: authorizationCode,
+                grant_type: "authorization_code",
+            }),
+            headers: { "Content-Type": "application/x-www-form-urlencoded" },
+        }
+    );
```
Comment (navarroaxel) [inline posts/nachoSource/01.revoking-apple-token.md:141]:
```suggestion
    );
    
    const {refresh_token} = await authTokenResponse.json();
```

File: posts/nachoSource/01.revoking-apple-token.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+    return await fetch("https://appleid.apple.com/auth/revoke", {
+        method: "POST",
+        body: JSON.stringify({
+            client_id: parameter.clientId,
+            client_secret: clientSecret,
+            token: authTokenResponse.data.refresh_token,
```
Comment (navarroaxel) [inline posts/nachoSource/01.revoking-apple-token.md:148]:
```suggestion
            token: refresh_token,
```

File: posts/nachoSource/01.revoking-apple-token.md [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+    return fetch('https://appleid.apple.com/auth/revoke', {
+        method: "POST",
+        body: JSON.stringify({
+            client_id: YOUR_CLIENT_ID,
+            client_secret: clientSecret,
+            token: identityToken,
+        }),
+        headers: { "Content-Type": "application/x-www-form-urlencoded" },
+    });
```
Comment (navarroaxel) [inline posts/nachoSource/01.revoking-apple-token.md:115]:
```suggestion
    const response = await fetch('https://appleid.apple.com/auth/revoke', {
        method: "POST",
        body: JSON.stringify({
            client_id: YOUR_CLIENT_ID,
            client_secret: clientSecret,
            token: identityToken,
        }),
        headers: { "Content-Type": "application/x-www-form-urlencoded" },
    });
    
    return response.json()
```

**Review by pabloromeo — APPROVED** [review]

File: posts/nachoSource/01.revoking-apple-token.md [hunk 10]
```diff
@@ -0,0 +1,158 @@
+---
+title: Revoking Apple ID
+published: false
+description: Revoke Apple ID in order to pass app versions review for iOS
+tags: 'ios, mobile, javascript, reactnative'
+cover_image: ./assets/app-store.jpeg
```
Comment (pabloromeo) [inline posts/nachoSource/01.revoking-apple-token.md:6]:
```suggestion
cover_image: ./assets/app-store.png
```

**Review by pabloromeo — CHANGES_REQUESTED** [review]

File: posts/nachoSource/01.revoking-apple-token.md [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+  );
+};
+```
+
+## 3. Consuming the API
+
+At last, calling this API returns a _200_ code if the previous steps were fine. If not, the response will be a _400_ (bad request) with some info.
+
+```javascript
+const revokeToken = (identityToken) => {
```
Comment (navarroaxel) [inline posts/nachoSource/01.revoking-apple-token.md:104]:
```suggestion
const revokeToken = async (identityToken) => {
```

**Review by pabloromeo — APPROVED** [review]

**Pull request closed without merging**